### ISSUE TITLE
Issues with macOS after commit dead206

### ISSUE THREAD

**Issue by AirPort** [mpv-player/mpv#4615]
### mpv version and platform

git master
macOS 10.12.5

### Reproduction steps

Play a video setting alternatively the `opengl-pbo` and the `opengl-early-flush=no` options

### Results

1. With `opengl-early-flush=no` ("needed" to play hi-fps video in audio-sync mode), the image flickers a lot both going on and off fullscreen if the video is bigger in at least one dimension than the screen resolution, everytime.

2. `opengl-pbo` also causes flicker going on and off fullscreen, with any kind of video but in a less "reliable" way (roughly half the times I tested, I think).

3. Before dead206, the `opengl-pbo` option used to help a lot with display-resample in fullscreen: with the option on, vsync-jitter value was pretty much stable at ~0.03, with no dropped frames and few missed ones. After the commit, the vsync-jitter value is 10x or more, and the framedrop counter runs continuosly.

This with both GPUs I can test on (integrated intel and discrete AMD).

~(It's worth noting that incidentally the commit had ripercussion on the flickering system indicator/dock issue in https://github.com/mpv-player/mpv/issues/3746 on my machine: now they don't flicker anymore out the box, and setting the opengl-pbo option that used to be a workaround now actually makes them flicker).~ 
EDIT: Nope, turns out it just flicker at random, and without workarounds now.

### Log file

The verbose logs don't seem to show any difference before/after that commit. If you need stats dumps or other things I should test let me know.

**Comment by Akemi** [comment]
i am wondering if adding double buffering back will fix it. can you try this [branch/commit](https://github.com/Akemi/mpv/commit/7bdac82f5c051323d4ed33eafbc2f1a0e9f15cef), pls.

**Comment by AirPort** [comment]
It indeed seems to fix the aforementioned issues, but on the other hand it seems to have a noticeable impact on performaces in fs, especially with display-resample and hi-fps videos (the old problem started with El Capitan?).

**Comment by Akemi** [comment]
yeah most likely the old problem. i kinda hoped that it would be magically fixed now.

if you remember my little [CAOpenGLLayer test](https://github.com/mpv-player/mpv/pull/4391#issuecomment-298945776) from PR #4391, would you mind testing it again with libmpv from current master?

**Comment by Akemi** [comment]
if you use stats.lua it can be shown via `echo 'script-binding stats/display_stats_toggle' | socat - /tmp/mpvsocket` in a separate terminal, providing you have socat installed.

**Comment by AirPort** [comment]
Finally I can bring some good news: I used my good old 60 fps clip to test and unlike last time it seems to be smooth, both in fs and windowed.
The only weird thing is that sometimes when I start it, it begins to judder for some seconds, then it "gain" sync and it stays smooth (but maybe that's the limitation of the test program you mention in your original post?).

**Comment by haasn** [comment]
A logfile with `--msg-level=vo=debug` would be helpful.

**Comment by AirPort** [comment]
Neither option: https://0x0.st/DsP.txt
Both options: https://0x0.st/DsZ.txt
`opengl-pbo`: https://0x0.st/DsN.txt
`opengl-early-flush=no`: https://0x0.st/Dsq.txt
plain display-resample: https://0x0.st/Dsc.txt
display-resample + pbo: https://0x0.st/Dsb.txt

**Comment by Akemi** [comment]
i've got some time today and could test various things too. though i can't confirm every issue from @AirPort, since in the first place i never experienced the Dock/Volume overlay flicker or the fullscreen regression.

with current master i do see a slight performance regression. moving the window breaks display resample, or rather the estimated display fps drop quite drastically. which didn't happen prior to dead206. i do see a flicker every now and then, but it isn't very drastic. `opengl-pbo` doesn't make a difference. vsync jitter stays between 0.032-0.04. before it usually was <0.032.

current master + re-adding double buffering completely breaks the fullscreen performance for me now. i didn't have that problem before and never could reproduce that issue (#2392). vsync jitter rises up to ~1.0 and stays there for a few seconds and descend to >0.031. this repeats itself. in general with double buffering it performs even worse than without.

with my opengl-cb + CAOpenGLLayer test i get the best result. i don't experience any kind of issues. vsync jitter stays mostly constantly under 0.03. moving the window doesn't break display resample. i see no flicker.

> Finally I can bring some good news: I used my good old 60 fps clip to test and unlike last time it seems to be smooth, both in fs and windowed.
> The only weird thing is that sometimes when I start it, it begins to judder for a bit, matter of a few seconds, then it "gain" sync and it stays smooth (but maybe that's the limitation of the test program you mention in your original post?).

yeah there are few small limitations that could potentially cause this. though most likely the display link callback is not completely stable sometimes directly after initialising it.

**Comment by haasn** [comment]
Moving.. the window? Commit dead206 only changes the way textures are uploaded, it should have nothing to do with the window location.

Also if `opengl-pbo` doesn't make a difference, how can that commit be related if it only affects `opengl-pbo`?

**Comment by Akemi** [comment]
> Moving.. the window? Commit dead206 only changes the way textures are uploaded, it should have nothing to do with the window location.

yeah moving the window. i am just reporting what i am experience, and i am not going to claim that it makes sense. who knows what macOS and its window manager is doing but it's joking on it when moving the window around. probably just because the performance in general took a hit.

> Also if opengl-pbo doesn't make a difference, how can that commit be related if it only affects opengl-pbo?

to make sure i tested dead206 and the commit before it 46bfa37. dead206 both `opengl-pbo=yes` and `opengl-pbo=no` have the window moving problem. on 46bfa37 `opengl-pbo=yes` doesn't have the problem but the problem occurs with `opengl-pbo=no`. since i have pbo activated in my config i just didn't notice. so yeah it makes sense, just that since dead206 `opengl-pbo=yes` also has that problem. also on dead206 `opengl-pbo=no` is worse than with `opengl-pbo=yes`.

**Comment by haasn** [comment]
Okay, so it seems like directly mapping the texture would be way better on OS X.

Possible solutions

1. Test/implement direct rendering as an alternative to mapped PBOs, and rely on that to solve OS X problems
2. Add a fallback path for mapped PBOs again and expand `--opengl-pbo` to allow you to choose any of the three paths

I'm planning on continuing work on #1 anyway, so it might be a good idea to just postpone this discussion until we have a better idea of whether or not DR helps.

**Comment by AirPort** [comment]
Now DR landed on master, but requires at least OpenGL 4.4: is there any chance for a try with possible solution #2?

**Comment by haasn** [comment]
Maybe. This stuff is actually handled by ra_gl instead of vo_opengl now so getting options into it might be harder.

**Comment by wm4** [comment]
>Before dead206, the opengl-pbo option used to help a lot with display-resample in fullscreen: with the option on, vsync-jitter value was pretty much stable at ~0.03, with no dropped frames and few missed ones. After the commit, the vsync-jitter value is 10x or more, and the framedrop counter runs continuosly.

That was most likely a coincidence, by causing the driver to flush or something (glFlush calls also affect our cocoa backend in ways it shouldn't). OSX GL display is in limbo (both in mpv, and apparently at Apple), so the real solution is much more complex. Maybe changing the cocoa code to use opengl-cb will help.

**Comment by Akemi** [comment]
so the past few weeks i worked on a [cocoa opengl-cb 'backend'](https://github.com/Akemi/mpv/commits/macOS_cocoa-cb2) that fixes all the annoying issues we still have, at least from my tests. i still need to clean-up the code, there are still some segfaults, i need to add some build related things, etc. there is still one big issue, the initialisation. it could happen that the core starts before the opengl context is set and mpv just quits. with a 'complex' enough config that won't happen and is enough for testing.

anyway, if you want to test the linked branch just set `--vo=opengl-cb`.

**Comment by AirPort** [comment]
Oh, I noticed some work in the test program repo some weeks ago, really happy to see it's coming along in "real" mpv. Just a question: the build exit with an error in the linking step; does it need the entire Xcode environment to build?

```
ld: warning: directory not found for option '-L/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/lib/swift_static/macosx'
ld: library not found for -lswiftCore for architecture x86_64

```

**Comment by Akemi** [comment]
yeah for now.

**Comment by AirPort** [comment]
Time to dig up my old machine then. Will report ASAP.

**Comment by Akemi** [comment]
seems like the equivalent command line tools directory would be `/Library/Developer/CommandLineTools/usr/lib/swift_static/macosx`.

**Comment by AirPort** [comment]
Yep, changed the line manually in wscript_build and it worked.
I did just a few quick tests and the results are really good, and pretty much any of the annoying issues seems to have disappeared: system overlays doesn't flicker anymore as far as I can see, same for the fs animation. I didn't notice any black flicker on resize, and performances are good in both windowed and fs, no matter audio or video-sync.
Just a few minor notes (although you probably already know all of this)

- The OSD looks blurry.
- Sometimes the window's title is cut after a few letters.
- The OSC has a strange behaviour, it appears only when the cursor is on a precise, little area of the player window.
- A couple of opengl options doesn't seem to work, like icc-profile-auto or the timers (probably fault of opengl-cb?).
- Sometines it crashes on exit.

(Here's a couple images to show the first two point of the list: [master](https://0x0.st/5F9.png) and [macOS_cocoa-cb2](https://0x0.st/5Fp.png))

**Comment by Akemi** [comment]
i was actually not aware of the first 3. i probably know why the first one occurs, it's related to being on a retina resolution. most likely easily fixable. for two no idea yet and 3 sounds weird but maybe related to retina again.

yeah icc-profile-auto  still shows that error that it isn't supported. i already implemented the support but there is still a segfault somewhere so i deactivated it. the crash is something easily fixable, i just put it at the end of my TODO list.

**Comment by Akemi** [comment]
oh i forgot the timers. yeah that is because of opengl-cb. no idea if this can be added to opengl-cb. it's probably up to @haasn, if at all. @Argon- probably would like it work for his stats.lua

**Comment by wm4** [comment]
Possibly you're creating a legacy GL context.

**Comment by Akemi** [comment]
to which issue might this be related?

the blurriness is because i use the [pt values](https://github.com/Akemi/mpv/blob/9292e7cfa1e4538b6a77f38729fb9cd5fe067591/video/out/cocoa-cb/video_layer.swift#L52) instead of actual pixels when drawing.

**Comment by wm4** [comment]
Missing timers, I was thinking.

**Comment by Akemi** [comment]
oh those timers are supposed to work with opengl-cb (eg the second page of stats.lua)? according to the log it detects the normal desktop opengl 4.1 like it is supposed to do. @haasn also reverted (#4721) his change, so it should also work on macOS again.

**Comment by wm4** [comment]
Also the perf voctrl is missing, which could break this too.

**Comment by haasn** [comment]
The perf VOCTRL is required for stats.lua. (The timers themselves should still be working without the voctrl, which you can check with `msg-level=vo=debug`)

**Comment by Akemi** [comment]
thx for the hint. that worked somewhat, besides the little [issue](https://0x0.st/RzG.mov) that fresh and redraw statistics are not returned properly it seems.

**Comment by Akemi** [comment]
@AirPort i fixed your reported issue 1, 4 (icc one, message still appears though, timers work now too though there is another bug see my previous post) and 5 (most likely). i couldn't reproduce 2 and i am not sure about 3. maybe you could explain the weird OSC behaviour a bit more.

**Comment by AirPort** [comment]
Retested:

- The blurry OSD is still [present](https://0x0.st/RXN.png).
- I can't reproduce your issue with timers: they seems to work fine here now, or at least like before (don't look at redraw's values in the screenshot, it has only showed me zeroes since the very first timers were introduced), and I also confirm that it couldn't make it crash even once on exit anymore.
- For the OSC thing, it seems to be a coordinates issue, as I noticed that I can interact with it by clicking with the mouse at the bottom of the upper left quadrant of the window. [Here's](https://0x0.st/RXA.png) a screenshot in which I highlighted a bunch of elements and the portion in the screen I need to click to trigger them. (the same issue is still present with --hidpi-window-scale=no)

**Comment by Akemi** [comment]
so something is really weird on your end. the title is cut off around the middle of the title bar. i don't have the coordinate issue, even on a retina resolution. both issues seem to be related as is the blurriness. which is also fixed here.

for me the timers report proper values.

**Comment by AirPort** [comment]
I noticed that the terminal says something I don't really get about geometry,  maybe it's useful to you so [here's](https://0x0.st/RXv.txt) a verbose log.

**Comment by Akemi** [comment]
okay i think i fixed the blurriness now for good. as dumb as it sounds one of the [layer events](https://github.com/Akemi/mpv/commit/862a94b65c3e68b945b62ee0ed21278c94e81099#diff-b5010a29091b72c707c97eae67a97254R131), that sets the content scale, doesn't trigger on init when using a window with border. since i use `no-border` the event triggered and i didn't see that bug.

this also fixes the mouse coordinates. the mpv core thought the viewport was only X by Y pixel wide and rendered at that resolution, since the content scale wasn't set. that led to mouse positions being clamped to X by Y even though the visible content layer was 2*X by 2*Y. the layer did the scaling to double that resolution, resulting in that 'offset' behaviour.

if you can confirm this fixed, that would leave only the title being cut off issue.

**Comment by AirPort** [comment]
Yep, it works (both the blurriness and the mouse coordinates)!

**Comment by Akemi** [comment]
for the title issue, could you try commenting [that line](https://github.com/Akemi/mpv/blob/macOS_cocoa-cb2/video/out/cocoa-cb_common.swift#L407) and replace it with a static title like `ccb.window.title = "A VERY LONG STRING FOR TESTING"`, and see if it works without cutting off?

**Comment by AirPort** [comment]
Nope, it get cut. Maybe interestingly enough, If I fullscreen and then go back to windowed the title is then complete.

**Comment by Akemi** [comment]
is it the same when just resizing the window?

**Comment by AirPort** [comment]
Yes. And I tried to open a 1080p file, and this too doesn't cut the title, so it seems it's a matter with the window's dimension.

(Another little thing I noticed: when this gets merged the default for `--opengl-early-flush` should be set to no for macOS too, because yes (which is the auto's pick for this platform, IIRC) totally destroys the performances with higher resolution videos).

**Comment by Akemi** [comment]
`--opengl-early-flush` actually still influence this? i thought it wouldn't with this, at least it doesn't here.

that title issue is very weird. i don't really do anything special and the titlebar is never touched. it kinda looks like a macOS bug tbh. just for testing if you use a none retina resolution does it still happen?

**Comment by AirPort** [comment]
>--opengl-early-flush actually still influence this? i thought it wouldn't with this, at least it doesn't here.

At least on weaker GPUs it does as far as I can see, a lot. On my machine I can't notice much of a difference using the dGPU, but with the iGPU `--opengl-early-flush=no` produces a seemingly normal playback, with `--opengl-early-flush=yes` it's unwatchably choppy (at least on >720p videos).

>that title issue is very weird. i don't really do anything special and the titlebar is never touched. it kinda looks like a macOS bug tbh. just for testing if you use a none retina resolution does it still happen?
It does with every resolution I can test on this machine.
As for the macOS bug, could be. I actually have a testing machine with High Sierra but it doesn't have a retina display unfortunately.

**Comment by Akemi** [comment]
i added [this](https://github.com/Akemi/mpv/commit/a768e5ef80a799a400fc2a31d4ec0377f365e4ed) to my branch as a test. does it make the title appear properly?

otherwise i am wondering if you change the hardcoded standard title [here](https://github.com/Akemi/mpv/blob/a768e5ef80a799a400fc2a31d4ec0377f365e4ed/video/out/cocoa-cb/window.swift#L42) to something longer, does it cut of less of the proper title?

**Comment by AirPort** [comment]
>i added this to my branch as a test. does it make the title appear properly?

Nope, same as before

>otherwise i am wondering if you change the hardcoded standard title here to something longer, does it cut of less of the proper title?

Well yes, it does. I tinkered with it a bit, and discovered that if I harcode in that line a string with a number of characters equal or superior to the video title, this shows up regularly. Anything less and it's cut, to various degrees.

**Comment by Akemi** [comment]
so it somehow doesn't update the bounds from the old title when the new one is set, cutting it off in the process. very weird.

**Comment by Akemi** [comment]
i updated my branch with a possible fix. does this fix it?

**Comment by AirPort** [comment]
No differences, sorry.

**Comment by Akemi** [comment]
i pushed some changes that print some things to the terminal. mind rebuilding and pasting me the output of the terminal pls?

**Comment by AirPort** [comment]
Sure, [here](https://0x0.st/Rc7.txt).

**Comment by Akemi** [comment]
okay, that is the exact same as here. i pushed another change to my branch. if that fixes it i was just being an idiot.

**Comment by AirPort** [comment]
Yep, it seems that fixed it!

**Comment by Akemi** [comment]
okay, that's great. i forgot that i need to update UI stuff on the main thread. this kinda explains the undefined behaviour. sry for all the trouble.

**Comment by AirPort** [comment]
Np, I'm happy to help where I can.

**Comment by Akemi** [comment]
fixed via c5e4538bc4b63f02bf89f62aa25a37b9eb0c0316.
